editor/plugins/area_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,158 @@
+		Vector2 mposxd = xform.affine_inverse().xform(mpos);
+
+		real_t grab_threshold = EDITOR_GET("editors/polygon_editor/point_grab_radius");
+		grab_threshold = grab_threshold * grab_threshold;
+
+		// Check for gravity point movement start.
+		if (action == ACTION_NONE && mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT) {
+			real_t dist_to_p = mpos.distance_squared_to(point);
+			if (dist_to_p < grab_threshold) {
+				action = ACTION_MOVING_POINT;
+				moving_from = node->get_gravity_point_center();
+				moving_mouse_from = mposxd;
+
+				canvas_item_editor->update_viewport();
+				return true;
+			}
+		}
+
+		// Check for gravity point movement completion.
+		if (action == ACTION_MOVING_POINT && !mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT) {
+			EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+			Vector2 new_pos = moving_from + mposxd - moving_mouse_from;
+
+			undo_redo->create_action(TTR("Move Gravity Point Center"));
+			undo_redo->add_do_method(node, "set_gravity_point_center", new_pos);
+			undo_redo->add_undo_method(node, "set_gravity_point_center", moving_from);
+			undo_redo->add_do_method(canvas_item_editor, "update_viewport");
+			undo_redo->add_undo_method(canvas_item_editor, "update_viewport");
+			undo_redo->commit_action();
+
+			action = ACTION_NONE;
+			return true;
+		}
+	}
+
+	// Handle movement.
+	Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid() && action == ACTION_MOVING_POINT) {
+		Transform2D xform = canvas_item_editor->get_canvas_transform() * node->get_global_transform();
+		Vector2 new_pos = moving_from + xform.affine_inverse().xform(mm->get_position()) - moving_mouse_from;
+
+		node->set_gravity_point_center(new_pos);
+
+		canvas_item_editor->update_viewport();
+		return true;
+	}
+
+	return false;
+}
+
+void Area2DEditorPlugin::forward_canvas_draw_over_viewport(Control *p_overlay) {
+	if (!node || !node->is_visible_in_tree() || !node->is_gravity_a_point() || node->get_gravity_space_override_mode() == Area2D::SpaceOverride::SPACE_OVERRIDE_DISABLED) {
+		return;
+	}
+
+	Transform2D xform = canvas_item_editor->get_canvas_transform() * node->get_global_transform();
+	const Ref<Texture2D> gravity_point_handle = canvas_item_editor->get_editor_theme_icon(SNAME("EditorHandle"));
+	const Size2 handle_size = gravity_point_handle->get_size();
+	Vector2 point = xform.xform(node->get_gravity_point_center());
+
+	if (node->get_gravity_point_unit_distance() > 0) {
+		float gdist = xform.get_scale().x * node->get_gravity_point_unit_distance();
+		p_overlay->draw_circle(point, gdist, Color(0.92f, 1.0f, 0.6f, 0.3f));
+	} else {
+		p_overlay->draw_circle(point, 20.0f, Color(0.5f, 0.5f, 0.5f, 0.3f));
+	}
+
+	p_overlay->draw_texture_rect(gravity_point_handle, Rect2(point - handle_size * 0.5, handle_size));
+}
+
+void Area2DEditorPlugin::_property_edited(const StringName &p_prop) {
+	if (((String)p_prop).begins_with("gravity")) {
```

Comment:
Is this guaranteed to not catch any unrelated properties?

Reply:
It's not, it can even catch user properties.
Although it's not a big problem. In worst case it causes unnecessary viewport refresh when Area2D is selected.

Reply:
True it does just refresh, but if the behavior here changes it'd need to be accounted for, but then we'd check the specific properties anyway to update handles etc.

Reply:
Though I'd suspect that the performance overhead of checking multiple strings is less than the refreshing at bad times

Reply:
We usually try to match as much of the string as possible to prevent false positives.
```suggestion
	if (((String)p_prop).begins_with("gravity_")) {
```